File: DukeMapGenLib/src/java/trn/duke/experiments/WallPrefab.java
package trn.duke.experiments;

import trn.Wall;

/**
 * Spec that describes a wall to create, to make it easier to create multiple walls.
 * 
 * 
 * @author Dave
 *
 */
public class WallPrefab {

	Integer texture = null;
	Short xrepeat = null;
	Short yrepeat = null;
	Short shade = null;
	
	public WallPrefab(Integer texture){
		this.texture = texture;
		this.xrepeat = 16;
		this.yrepeat = 8;
	}
	
	public WallPrefab(WallPrefab copyMe){
		this.texture = copyMe.texture;
		this.xrepeat = copyMe.xrepeat;
		this.yrepeat = copyMe.yrepeat;
		this.shade = copyMe.shade;
	}
	
	public WallPrefab setTexture(int texture){
		this.texture = texture;
		return this;
	}
	
	public WallPrefab setXRepeat(int xrepeat){
		this.xrepeat = (short)xrepeat;
		return this;
	}
	
	public WallPrefab setYRepeat(int yrepeat){
		this.yrepeat = (short)yrepeat;
		return this;
	}
	
	public WallPrefab setShade(short shade){
		this.shade = shade;
		return this;
	}
	
	public void writeTo(Wall... walls){
		for(Wall w : walls){
			if(texture != null){
				w.setTexture(texture);
			}
			
			if(xrepeat != null){
				w.setXRepeat(xrepeat);
			}
			
			if(yrepeat != null){
				w.setYRepeat(yrepeat);
			}
			
			if(shade != null){
				w.setShade(shade);
			}	
		}
		
	}
	
	
	
	
}

File: DukeMapGenLib/src/scala/duchy/experiments/render/maze/WallPrefab.scala
package duchy.experiments.render.maze

import trn.{WallBrush, Wall}

object WallPrefab {

  def apply(tex: Int): WallPrefab = WallPrefab(
    texture=tex,
    xrepeat=Some(16),
    yrepeat=Some(8),
    shade=None,
  )

  def apply(copyMe: WallPrefab): WallPrefab = copyMe.copy()

}

/**
  * Legacy class
  *
  * TODO there is a newer one at trn.render.WallPrefab
  */
case class WallPrefab (
  texture: Integer,
  var xrepeat: Option[Short],
  var yrepeat: Option[Short],
  var shade: Option[Short],
) extends WallBrush {

  override def writeToWall(w: Wall): Unit = {
    w.setTexture(texture)
    xrepeat.foreach(w.setXRepeat)
    yrepeat.foreach(w.setYRepeat)
    shade.foreach(w.setShade)
  }

  def withTexture(tex: Int): WallPrefab = this.copy(texture=tex)

  def setXRepeat(xrepeat: Int): WallPrefab = {
    this.xrepeat = Some(xrepeat.toShort)
    this
  }

  def setYRepeat(yrepeat: Int): WallPrefab = {
    this.yrepeat = Some(yrepeat.toShort)
    this
  }

  def setShade(shade: Short): WallPrefab = {
    this.shade = Some(shade)
    this
  }

  def writeToWalls(walls: Wall*): Unit = {
    for (w <- walls) {
      writeToWall(w)
    }
  }
}
